zend_abstract_interface/functions/tests/CMakeLists.txt

```diff
@@ -0,0 +1,11 @@
+#[[ TODO When we have a consistent API across all PHP versions with ZAI
+    data-structure shims, move these tests out of PHP-versioned directories.
+ ]]
+add_executable(functions "${PHP_VERSION_DIRECTORY}/functions.cc")
+
+target_link_libraries(functions PUBLIC catch2_main Zai::Sapi Zai::Functions)
+
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/stubs
+     DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
```

Comment:
I understand why the copy is necessary, but I don't like it for some reason I can't place. Did you have a similar feeling or is it just me?

Reply:
The only quirky thing that I've experienced with this copy is that if you make a change to a stub, you basically have to blow away the build dir and start over again (at least that's what I do with my primitive `cmake` skills. 😆) But I'm not sure how else we'd get the stubs copied over for the tests without doing a `COPY`. Of course I'm totally open to ideas if you've got a better solution!

Reply:
I think you need to add the copied file as a dependency of the test using `add_dependency`, though I'm not sure a `COPY` is recognized as a "factory" (I use quotes because I'm not sure what the proper CMake terminology is).

Reply:
As far as I can tell from the [`add_dependencies` docs](https://cmake.org/cmake/help/latest/command/add_dependencies.html) (I couldn't find `add_dependency`), the API only supports targets to resolve build-order dependencies. I converted the stubs dir to a target with `add_custom_target`, but this did not copy the files:

```
add_custom_target(function_stubs SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/stubs)
add_dependencies(functions function_stubs)
```

I was able to copy the files using [`add_custom_command`](https://cmake.org/cmake/help/latest/command/add_custom_command.html):

```
add_custom_command(TARGET functions
                   POST_BUILD
                   COMMAND ${CMAKE_COMMAND} -E copy_directory
                           ${CMAKE_CURRENT_SOURCE_DIR}/stubs
                           ${CMAKE_CURRENT_BINARY_DIR}/stubs)
```

However, this still will not copy the stub files when they are changed. It will only recopy if you make a change to a source file from the target. Effectively it's the same behavior as `COPY`.

There might be a better solution out there, but I haven't been able to find it.

Reply:
We'll just leave this alone for now, as the proper way eludes us.